volttron/platform/agent/base_historian.py

```diff
@@ -1540,19 +1540,25 @@ def remove_successfully_published(self, successful_publishes,
             else:
                 self._record_count -= c.rowcount
         else:
-            temp = list(successful_publishes)
-            temp.sort()
+            # translate topics to topic_ids
+            updated_succsessful_publishes = [{'timestamp': x[0], 'topic_id': self._backup_cache[x[1]]}
+                                             for x in successful_publishes]
             c.executemany('''DELETE FROM outstanding
-                            WHERE id = ?''',
-                          ((_id,) for _id in
-                           successful_publishes))
-            self._record_count -= len(temp)
+                            WHERE ts = ? AND topic_id = ?''',
+                          ((d['timestamp'], d['topic_id']) for d in
+                           updated_succsessful_publishes))
+
+            # rowcount gives the number of rows that were actually deleted in the cache
+            # see https://docs.python.org/3/library/sqlite3.html#sqlite3.Cursor.rowcount
+            self._record_count -= c.rowcount
 
+        _log.debug(f"Count of rows in cache after deletion: {self._record_count}")
         self._connection.commit()
 
     def get_outstanding_to_publish(self, size_limit):
         """
-        Retrieve up to `size_limit` records from the cache.
+        Retrieve up to `size_limit` records from the cache. Guarantees a unique list of records,
+        where unique is defined as topic + timestamp.
```

Comment:
topic + timestamp is assuming concatenation...use tuple syntax (topic, timestamp) .